### ISSUE TITLE
"Rollbar already initialized. Ignoring re-init." for test cases involving Client()

### ISSUE THREAD

**Issue by smason** [rollbar/pyrollbar#145]
I'm just getting going on testing my Django code and I'm getting lots of:
```
.WARNING:rollbar:Rollbar already initialized. Ignoring re-init.
.WARNING:rollbar:Rollbar already initialized. Ignoring re-init.
```
messages—almost one for every test that involves a `Client()`, but seems to be less sometimes.

Pretty sure this doesn't relate to #100 as I'm not calling `rollbar.init()` anywhere.  A `grep` for rollbar only shows appropriate looking `import`s and calls to `rollbar.report_message`.

In case it matters, I'm using Django (1.10.3) and rollbar (0.13.8) under Python 2.7.12.

**Comment by ezarowny** [comment]
Interesting! Is this causing any problems for you right now?

**Comment by smason** [comment]
@ezarowny Yes, currently doing this every time I run my tests.  Somewhat annoying, but not a show-stopper

**Comment by smason** [comment]
Have just tried putting a stack dump in and it seems to always come out with a similar dump.  I'll annotate as I understand it:
```
  File "/usr/lib64/python2.7/unittest/case.py", line 393, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib64/python2.7/unittest/case.py", line 329, in run
    testMethod()
```
the above are the final couple of calls setting up my `django.test.TestCase`, there are quite a few lines before it but am assuming they are of little interest:
```
  File "/home/smason/work/concon/consultant-connect/src/api/tests.py", line 11, in test_build_positive
    response = Client().get('/api/v1/app-version-check/', {'buildnum':-50})
```
OK, my code is being run and this line is instantiating a new `Client`, i.e. from `django.test.Client`.  This cases internal stuff of the Client to run:
```
  File "/home/smason/.virtualenvs/concon/lib/python2.7/site-packages/django/test/client.py", line 529, in get
    **extra)
  File "/home/smason/.virtualenvs/concon/lib/python2.7/site-packages/django/test/client.py", line 333, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/home/smason/.virtualenvs/concon/lib/python2.7/site-packages/django/test/client.py", line 409, in generic
    return self.request(**r)
  File "/home/smason/.virtualenvs/concon/lib/python2.7/site-packages/django/test/client.py", line 476, in request
    response = self.handler(environ)
```
until it tries reloading the django middleware, at which point your code complains:
```
  File "/home/smason/.virtualenvs/concon/lib/python2.7/site-packages/django/test/client.py", line 129, in __call__
    self.load_middleware()
  File "/home/smason/.virtualenvs/concon/lib/python2.7/site-packages/django/core/handlers/base.py", line 82, in load_middleware
    mw_instance = middleware(handler)
  File "/home/smason/.virtualenvs/concon/lib/python2.7/site-packages/rollbar/contrib/django/middleware.py", line 101, in __init__
    rollbar.init(access_token, environment, **kw)
  File "/home/smason/.virtualenvs/concon/lib/python2.7/site-packages/rollbar/__init__.py", line 278, in init
    traceback.print_stack()
WARNING:rollbar:Rollbar already initialized. Ignoring re-init.
```

So it seem my hypothesis above was somewhat correct, you should see an error for most `Client`s that do some sort of request.

**Comment by ezarowny** [comment]
Does this happen if you disable the Rollbar middleware via settings?

**Comment by smason** [comment]
@ezarowny, Yes.

Is it expected that one should remove Rollbar from middleware when doing tests?  I've not noticed anything in the docs to that effect, but is certainly easy to do.

**Comment by smason** [comment]
Assuming that my expectation is correct, I have just found `django.test.modify_settings` which seems very useful for this sort of thing.  I've created some base classes (as it doesn't seem to play with Mixins) like:
```python
from django import test
@test.modify_settings(MIDDLEWARE={'remove':[
    'rollbar.contrib.django.middleware.RollbarNotifierMiddleware',
]})
class TestCase(test.TestCase):
    pass
```
and similar for the `LiveServerTestCase`, then derive all my tests from these classes instead.  Hope this is useful for others!

**Comment by ezarowny** [comment]
I'm not sure if it's expected that one should do that. I'll think on it and discuss with some other Rollbar engineers. Generally, one does not want external services being hit during unit tests though.

For what it's worth, I've removed the Rollbar Django middleware before. I had settings_test.py that inherited from settings.py.

**Comment by danidee10** [comment]
Just thought i should drop this here. The warnings don't appear in `Django==1.11`.

I think it has to do with suppressing Deprecation warnings in this release https://docs.djangoproject.com/en/1.11/releases/1.11/#deprecating-warnings-are-no-longer-loud-by-default

I'm not so sure because the docs say `DeprecationWarning` not all/third-party warnings. but at least you don't get to see those warning messages again 😄

**Comment by rivkahstandig3636** [comment]
Looks like this was solved with #198
